modules/raft-api/src/main/java/org/apache/ignite/internal/raft/rebalance/ChangePeersAndLearnersWithRetry.java

```diff
@@ -0,0 +1,102 @@
+import org.apache.ignite.internal.raft.service.RaftGroupService;
+import org.apache.ignite.internal.util.IgniteBusyLock;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Helper class that executes change peers and learners async with retries.
+ */
+public class ChangePeersAndLearnersWithRetry {
+
+    private final RaftCommandWithRetry raftCommand;
+
+    private final Supplier<CompletableFuture<RaftGroupService>> raftGroupServiceSupplier;
+
+    /**
+     * Creates a new instance of ChangePeersAndLearnersWithRetry.
+     *
+     * @param busyLock The busy lock.
+     * @param rebalanceScheduler The scheduler for rebalance tasks.
+     * @param raftGroupServiceSupplier The supplier of raft group service.
+     */
+    public ChangePeersAndLearnersWithRetry(
+            IgniteBusyLock busyLock,
+            ScheduledExecutorService rebalanceScheduler,
+            Supplier<CompletableFuture<RaftGroupService>> raftGroupServiceSupplier
+    ) {
+        this.raftGroupServiceSupplier = raftGroupServiceSupplier;
+
+        raftCommand = new RaftCommandWithRetry(busyLock, rebalanceScheduler);
+    }
+
+    /**
+     * Performs {@link RaftGroupService#changePeersAndLearnersAsync} on a provided raft group service of a partition, so nodes of the
+     * corresponding raft group can be reconfigured. Retry mechanism is applied to repeat
+     * {@link RaftGroupService#changePeersAndLearnersAsync} if previous one failed with some exception.
+     *
+     * @return Function which performs {@link RaftGroupService#changePeersAndLearnersAsync}.
+     */
+    public CompletableFuture<Void> execute(
+            PeersAndLearners peersAndLearners,
+            long sequenceToken,
+            Function<RaftGroupService, CompletableFuture<@Nullable RaftWithTerm>> leaderFilter) {
+
+        return raftCommand.execute(() ->
+                raftGroupServiceSupplier
+                        .get()
+                        .thenCompose(leaderFilter)
+                        .thenCompose(raftWithTerm -> {
+                            if (raftWithTerm == null) {
+                                return nullCompletedFuture();
+                            }
+
+                            return raftWithTerm.raftClient()
+                                    .changePeersAndLearnersAsync(peersAndLearners, raftWithTerm.term(), sequenceToken);
+                        }));
+    }
+
+    /**
+     * Performs {@link RaftGroupService#changePeersAndLearnersAsync} on a provided raft group service of a partition, so nodes of the
+     * corresponding raft group can be reconfigured. Retry mechanism is applied to repeat
+     * {@link RaftGroupService#changePeersAndLearnersAsync} if previous one failed with some exception.
+     *
+     * @return Function which performs {@link RaftGroupService#changePeersAndLearnersAsync}.
+     */
+    public CompletableFuture<Void> executeOnLeader(PeersAndLearners peersAndLearners, long term, long sequenceToken) {
```

Comment:
This method is very similar to a previous one. What do you think about reusing that one as
```
execute(peersAndLearners, sequenceToken, raftClient -> completedFuture(new RaftWithTerm(raftClient, term)));
```